programs/castle-vault/src/instructions/consolidate_refresh.rs

```diff
@@ -0,0 +1,186 @@
+#![allow(dead_code)]
+#![allow(unused_imports)]
+
+use boolinator::Boolinator;
+
+use anchor_lang::prelude::*;
+use anchor_spl::token::{self, Mint, MintTo, Token, TokenAccount};
+use port_anchor_adaptor::{port_lending_id, PortReserve};
+
+use crate::adapters::{solend, SolendReserve};
+use crate::errors::ErrorCode;
+use crate::reserves::Provider;
+use crate::state::{Vault, VaultFlags};
+use strum::IntoEnumIterator;
+
+// NOTE: having all accounts for each lending market reserve here is not scalable
+// since eventually we will hit into transaction size limits
+#[derive(Accounts)]
+pub struct ConsolidateRefresh<'info> {
+    /// Vault state account
+    /// Checks that the accounts passed in are correct
+    #[account(
+        mut,
+        has_one = vault_authority,
+        has_one = vault_reserve_token,
+        has_one = lp_token_mint,
+    )]
+    pub vault: Box<Account<'info, Vault>>,
+
+    /// Authority that the vault uses for lp token mints/burns and transfers to/from downstream assets
+    pub vault_authority: AccountInfo<'info>,
+
+    /// Token account for the vault's reserve tokens
+    pub vault_reserve_token: Box<Account<'info, TokenAccount>>,
+
+    /// Mint for the vault lp token
+    #[account(mut)]
+    pub lp_token_mint: Box<Account<'info, Mint>>,
+
+    pub token_program: Program<'info, Token>,
+
+    pub clock: Sysvar<'info, Clock>,
+}
+
+// TODO refactor refresh cpi calls into adapter pattern
+impl<'info> ConsolidateRefresh<'info> {
+    /// CpiContext for collecting fees by minting new vault lp tokens
+    #[cfg(feature = "fees")]
+    fn mint_to_context(
+        &self,
+        fee_receiver: &AccountInfo<'info>,
+    ) -> CpiContext<'_, '_, '_, 'info, MintTo<'info>> {
+        CpiContext::new(
+            self.token_program.to_account_info(),
+            MintTo {
+                mint: self.lp_token_mint.to_account_info(),
+                to: fee_receiver.clone(),
+                authority: self.vault_authority.clone(),
+            },
+        )
+    }
+}
+
+/// updates the vault total value, and collects fees
+pub fn handler<'info>(ctx: Context<'_, '_, '_, 'info, ConsolidateRefresh<'info>>) -> ProgramResult {
+    #[cfg(feature = "debug")]
+    msg!("Consolidate vault refreshing");
+
+    // Check that refreshes are not halted
+    (!ctx
+        .accounts
+        .vault
+        .flags()
+        .contains(VaultFlags::HALT_REFRESHES))
+    .ok_or::<ProgramError>(ErrorCode::HaltedVault.into())?;
+
+    // Calculate new vault value
+    let vault_reserve_token_amount = ctx.accounts.vault_reserve_token.amount;
+    let mut vault_value = vault_reserve_token_amount;
```

Comment:
This would be nicer using combinators:
```
    let vault_value =
        Provider::iter().try_fold(ctx.accounts.vault_reserve_token.amount, |acc, p| {
            let allocation = ctx.accounts.vault.actual_allocations[p];
            (allocation
                .last_update
                .slots_elapsed(ctx.accounts.clock.slot)?
                == 0)
                .as_result::<u64, ProgramError>(
                    acc.checked_add(allocation.value)
                        .ok_or(ErrorCode::OverflowError)?,
                    ErrorCode::AllocationIsNotUpdated.into(),
                )
        });
```

Reply:
missing a `?` at the end